### PR TITLE
Fea: Replace use system

### PR DESCRIPTION
**Pull request by mzegar**

### CODE REVIEW

**Comment by mzegar** [comment]
Let's merge https://github.com/TTTReborn/tttreborn/pull/345 first since I have that code here to get the game mode to compile.

File: code/items/ammo/TTTAmmo.cs [hunk 1]
```diff
@@ -50,58 +52,83 @@ public override void Spawn()
...
-            inventory.Ammo.Give(ammoType, amountGiven);
-            CurrentAmmo -= amountGiven;
-            OnPickup.Fire(other);
+        public float HintDistance => 80f;
 
-            if (CurrentAmmo <= 0 || Math.Ceiling(AmmoEntMax * 0.25) > CurrentAmmo)
-            {
-                Delete();
-            }
+        public TranslationData TextOnTick => new("GENERIC_PICKUP", new object[] { Input.GetKeyWithBinding("+iv_use").ToUpper() });
```
Comment (mzegar) [inline code/items/ammo/TTTAmmo.cs:77]: We need a way to provide the translated string for the item, is there a way to do this? We would need to use `GetTranslation` @Alf21

**Comment by mzegar** [comment]
Closes https://github.com/TTTReborn/tttreborn/issues/341
Closes https://github.com/TTTReborn/tttreborn/issues/332

File: code/items/ammo/TTTAmmo.cs (hunk 1 above)
Comment (Alf21) [reply code/items/ammo/TTTAmmo.cs:77]: Already possible <https://github.com/TTTReborn/tttreborn/blob/086d86f86f41b95cc5626a2ff699e17e018aac39/code/lang/Language.cs#L98-L101, e.g. used here <https://github.com/TTTReborn/tttreborn/blob/40bfd72fe15b23a90d1cfe91f5f9b60bf3d54e82/code/ui/generalhud/menu/Menu.ShopEditor.ItemEdit.cs#L37>

File: code/items/ammo/TTTAmmo.cs [hunk 2]
```diff
@@ -50,58 +52,83 @@ public override void Spawn()
...
+                Inventory inventory = player.Inventory;
+
+                if (!inventory.GetAmmoTypes().Contains(ammoType))
+                {
+                    return;
+                }
+
+                int playerAmount = inventory.Ammo.Count(ammoType);
+
+                if (!(Max >= (playerAmount + Math.Ceiling(CurrentAmmo * 0.25))))
```
Comment (Alf21) [inline code/items/ammo/TTTAmmo.cs:118]: lol dafuq

```suggestion
                if (Max < playerAmount + Math.Ceiling(CurrentAmmo * 0.25))
```

File: code/player/IEntityHint.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        /// <summary>
+        /// The hint we should display.
+        /// </summary>
+        public EntityHintPanel DisplayHint(TTTPlayer client);
+
+        /// <summary>
+        /// Occurs on each tick if the hint is active.
+        /// </summary>
+        public void Tick(TTTPlayer player);
+    }
```
Comment (Alf21) [inline code/player/IEntityHint.cs:39]:
```suggestion
    public interface IEntityHint
    {
        /// <summary>
        /// The max viewable distance of the hint.
        /// </summary>
        float HintDistance => 2048f;

        /// <summary>
        /// If we should show a glow around the entity.
        /// </summary>
        bool ShowGlow => true;

        /// <summary>
        /// The text to display on the hint each tick.
        /// </summary>
        TranslationData TextOnTick => null;

        /// <summary>
        /// Whether or not we can show the UI hint.
        /// </summary>
        bool CanHint(TTTPlayer client);

        /// <summary>
        /// The hint we should display.
        /// </summary>
        EntityHintPanel DisplayHint(TTTPlayer client);

        /// <summary>
        /// Occurs on each tick if the hint is active.
        /// </summary>
        void Tick(TTTPlayer player);
    }
```

File: code/player/PlayerCorpse.cs [hunk 4]
```diff
@@ -140,11 +141,56 @@ public ConfirmationData GetConfirmationData()
...
         public EntityHintPanel DisplayHint(TTTPlayer client)
         {
             return new Hint(TextOnTick);
         }
+
+        public void Tick(TTTPlayer confirmingPlayer)
+        {
+            using (Prediction.Off())
+            {
+
```
Comment (Alf21) [inline code/player/PlayerCorpse.cs:155]:
```suggestion
```

File: code/player/TTTPlayer.Hints.cs [hunk 5]
```diff
@@ -14,42 +14,61 @@ public partial class TTTPlayer
...
-            _currentHintPanel.Parent = HintDisplay.Instance;
-            _currentHintPanel.Enabled = true;
+            if (IsClient)
+            {
+                if (hint.ShowGlow && hint is ModelEntity model)
+                {
+                    model.GlowColor = Color.White;
+                    model.GlowActive = true;
+                }
+                _currentHintPanel = hint.DisplayHint(this);
```
Comment (Alf21) [inline code/player/TTTPlayer.Hints.cs:53]:
```suggestion

                _currentHintPanel = hint.DisplayHint(this);
```

File: code/player/TTTPlayer.Hints.cs [hunk 6]
```diff
@@ -14,42 +14,61 @@ public partial class TTTPlayer
...
         {
-            _currentHintPanel?.Delete(true);
-            _currentHintPanel = null;
+            if (IsClient)
+            {
+                if (_currentHint != null && _currentHint is ModelEntity model && model.IsValid())
+                {
+                    model.GlowActive = false;
+                }
+                _currentHintPanel?.Delete(true);
```
Comment (Alf21) [inline code/player/TTTPlayer.Hints.cs:69]:
```suggestion

                _currentHintPanel?.Delete(true);
```

File: code/player/TTTPlayer.Hints.cs [hunk 7]
```diff
@@ -14,42 +14,61 @@ public partial class TTTPlayer
...
             }
 
             DeleteHint();
 
-            _currentHintPanel = hint.DisplayHint(player);
-            _currentHintPanel.Parent = HintDisplay.Instance;
-            _currentHintPanel.Enabled = true;
+            if (IsClient)
+            {
+                if (hint.ShowGlow && hint is ModelEntity model)
```
Comment (Alf21) [inline code/player/TTTPlayer.Hints.cs:48]: Here, you don't check for a valid model but below you do

File: code/player/TTTPlayer.cs [hunk 8]
```diff
@@ -215,12 +212,15 @@ protected override void UseFail()
...
+            if (IsClient)
             {
                 return;
             }
 
-            base.StartTouch(other);
+            if (other is PickupTrigger)
+            {
+                StartTouch(other.Parent);
+            }
```
Comment (Alf21) [inline code/player/TTTPlayer.cs:223]: No base call anymore?
And no drop pickup prevention?

**Review by Alf21 — CHANGES_REQUESTED** [review]
Great and clean changes!

File: code/items/ammo/TTTAmmo.cs (hunk 2 above)
Comment (mzegar) [reply code/items/ammo/TTTAmmo.cs:118]: Just copying and pasting previous code, we can clean it up

File: code/player/TTTPlayer.cs (hunk 8 above)
Comment (mzegar) [reply code/player/TTTPlayer.cs:223]:
```
		public override void StartTouch( Entity other )
		{
			if ( IsClient ) return;

			if ( other is PickupTrigger )
			{
				StartTouch( other.Parent );
				return;
			}

			Inventory?.Add( other, Inventory.Active == null );
		}
```

This is what base looks like, we do not want to do `Inventory?.Add( other, Inventory.Active == null );` anymore. Maybe for things that aren't `TTTWeapons`? Thoughts?

File: code/player/PlayerCorpse.cs [hunk 9]
```diff
@@ -140,11 +141,56 @@ public ConfirmationData GetConfirmationData()
...
+                    if (DeadPlayer != null && DeadPlayer.IsValid())
+                    {
+                        DeadPlayer.IsConfirmed = true;
+                        DeadPlayer.CorpseConfirmer = confirmingPlayer;
+
+                        int credits = DeadPlayer.Credits;
+
+                        if (credits > 0)
+                        {
+                            DeadPlayer.Credits += credits;
```
[the remark this answers is not in the document]
Comment (mzegar) [reply code/player/PlayerCorpse.cs:180]: Good catch.

File: code/player/TTTPlayer.cs (hunk 8 above)
Comment (Alf21) [reply code/player/TTTPlayer.cs:223]: Hmmm... Maybe if it comes to that point, but not for now... Ty for the explaination

File: code/items/ammo/BuckshotAmmo.cs [hunk 10]
```diff
@@ -7,7 +7,7 @@ namespace TTTReborn.Items
     [Hammer.EditorModel("models/ammo/ammo_buckshot.vmdl")]
     partial class BuckshotAmmo : TTTAmmo
     {
-        public override string AmmoName => "buckshot";
+        public override string AmmoName => "ammo_buckshot";
```
Comment (Alf21) [inline code/items/ammo/BuckshotAmmo.cs:10]: Just delete it and populate it in TTTAmmo on creation based on the LibaryName (like we did in the other entities) if not already done or not duplicated.

File: code/items/weapons/NewtonLauncher.cs [hunk 11]
```diff
@@ -5,7 +5,7 @@
 namespace TTTReborn.Items
 {
     [Library("weapon_newtonlauncher")]
-    [Weapon(SlotType = SlotType.OffensiveEquipment, AmmoType = "pistol")]
+    [Weapon(SlotType = SlotType.OffensiveEquipment, AmmoType = "ammo_pistol")]
```
Comment (Alf21) [inline code/items/weapons/NewtonLauncher.cs:8]: can we use a type in that place? Would make much more sense imo.

File: code/items/weapons/TTTWeapon.cs [hunk 12]
```diff
@@ -20,11 +22,11 @@ public WeaponAttribute() : base()
     }
 
     [Hammer.Skip]
-    public abstract partial class TTTWeapon : BaseWeapon, ICarriableItem
+    public abstract partial class TTTWeapon : BaseWeapon, ICarriableItem, IEntityHint
     {
         public string LibraryName { get; }
         public SlotType SlotType { get; } = SlotType.Secondary;
-        public string AmmoType { get; } = "pistol";
+        public string AmmoType { get; } = "ammo_pistol";
```
Comment (Alf21) [inline code/items/weapons/TTTWeapon.cs:29]: Maybe a type would make more sense.

**Review by Alf21 — CHANGES_REQUESTED** [review]

File: code/items/weapons/NewtonLauncher.cs (hunk 11 above)
Comment (mzegar) [reply code/items/weapons/NewtonLauncher.cs:8]: I agree, can we do this in a different PR? This PR is getting a bit large. I can handle the changes. (you are suggesting using an enum right?)

Comment (mzegar) [reply code/items/weapons/NewtonLauncher.cs:8]: https://github.com/TTTReborn/tttreborn/pull/347

**Review by Alf21 — APPROVED** [review]

**Pull request merged**